File: lc/coreprep/implementation/dp_reference_implementation/unbounded_knapsack_patterns/RodCuttingMemo.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class RodCuttingMemo {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int l = in.nextInt();

        int[] prices = new int[n];
        for (int i = 0; i < n; i++) {
          prices[i] = in.nextInt();
        }

        int[] lengths = new int[n];
        for (int i = 0; i < n; i++) {
          lengths[i] = in.nextInt();
        }

        System.out.println(rodCutting(prices, lengths, l));
      }
    }
  }

  // O(nl) / O(nl)
  public static int rodCutting(int[] prices, int[] lengths, int length) {
    if (lengths.length == 0 || length == 0) {
      return 0;
    }

    Integer[][] dp = new Integer[lengths.length][length + 1];
    return rodCutting(prices, lengths, length, dp, 0);
  }

  private static int rodCutting(int[] prices, int[] lengths, int length,
                                Integer[][] dp, int currIdx) {
    if (currIdx >= lengths.length) {
      return 0;
    }

    if (dp[currIdx][length] == null) {
      int profit1 = 0;

      if (lengths[currIdx] <= length) {
        profit1 =
            prices[currIdx] +
            rodCutting(prices, lengths, length - lengths[currIdx], dp, currIdx);
      }

      int profit2 = rodCutting(prices, lengths, length, dp, currIdx + 1);

      dp[currIdx][length] = Math.max(profit1, profit2);
    }

    return dp[currIdx][length];
  }
}

File: lc/coreprep/implementation/dp_reference_implementation/unbounded_knapsack_patterns/rod_cutting_memo.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits>
#include <list>
#include <queue>
#include <random>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

int next_random(int low, int high) {
  random_device rd;
  mt19937 engine(rd());
  uniform_int_distribution<int> dist(low, high);

  return dist(engine);
}

int max_profit(const vector<int> &prices, const vector<int> &lengths,
               int length, vector<vector<int>> &dp, int curr_idx) {
  if (length == 0) {
    return 0;
  }

  if (curr_idx >= lengths.size()) {
    return 0;
  }

  if (dp[curr_idx][length] == -1) {
    int profit1 = 0;

    if (lengths[curr_idx] <= length) {
      profit1 =
          prices[curr_idx] +
          max_profit(prices, lengths, length - lengths[curr_idx], dp, curr_idx);
    }

    int profit2 = max_profit(prices, lengths, length, dp, curr_idx + 1);

    dp[curr_idx][length] = max(profit1, profit2);
  }

  return dp[curr_idx][length];
}

// O(nl) / O(nl)
int max_profit(const vector<int> &prices, const vector<int> &lengths,
               int length) {
  if (lengths.empty() || length == 0) {
    return 0;
  }

  vector<vector<int>> dp(lengths.size(), vector<int>(length + 1, -1));
  return max_profit(prices, lengths, length, dp, 0);
}

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, length;
  cin >> tt;

  while (tt--) {
    cin >> n >> length;

    vector<int> prices(n);
    for (int i = 0; i < n; i++) {
      cin >> prices[i];
    }

    vector<int> lengths(n);
    for (int i = 0; i < n; i++) {
      cin >> lengths[i];
    }

    cout << max_profit(prices, lengths, length) << "\n";
  }

  return 0;
}